### ISSUE TITLE
Error binding to target method exception when serializing specific classes

### ISSUE THREAD

**Issue by Siderite** [karlseguin/Metsys.Bson#1]
Hi, I wanted to test your library and I accidentally stumbled upon a pair of classes that cause the Serializer.Serialize method to bust. The problem appears on serializing CssClassValue.

 Couldn't find a way to link files, so here are the text sources and exception details:

========== Exception details =============

System.Reflection.TargetInvocationException was unhandled
  Message="Exception has been thrown by the target of an invocation."
  Source="mscorlib"
  StackTrace:
       at System.RuntimeMethodHandle._InvokeMethodFast(Object target, Object[] arguments, SignatureStruct& sig, MethodAttributes methodAttributes, RuntimeTypeHandle typeOwner)
       at System.RuntimeMethodHandle.InvokeMethodFast(Object target, Object[] arguments, Signature sig, MethodAttributes methodAttributes, RuntimeTypeHandle typeOwner)
       at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture, Boolean skipVisibilityChecks)
       at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
       at System.Reflection.MethodBase.Invoke(Object obj, Object[] parameters)
       at Metsys.Bson.MagicProperty.CreateGetterMethod(PropertyInfo property) in C:\Projects\karlseguin-Metsys.Bson-20b88f8\Metsys.Bson\Helpers\MagicProperty.cs:line 56
       at Metsys.Bson.MagicProperty..ctor(PropertyInfo property, String name, Boolean ignored, Boolean ignoredIfNull) in C:\Projects\karlseguin-Metsys.Bson-20b88f8\Metsys.Bson\Helpers\MagicProperty.cs:line 41
       at Metsys.Bson.TypeHelper.LoadMagicProperties(Type type, IEnumerable`1 properties) in C:\Projects\karlseguin-Metsys.Bson-20b88f8\Metsys.Bson\Helpers\TypeHelper.cs:line 101
       at Metsys.Bson.TypeHelper..ctor(Type type) in C:\Projects\karlseguin-Metsys.Bson-20b88f8\Metsys.Bson\Helpers\TypeHelper.cs:line 20
       at Metsys.Bson.TypeHelper.GetHelperForType(Type type) in C:\Projects\karlseguin-Metsys.Bson-20b88f8\Metsys.Bson\Helpers\TypeHelper.cs:line 44
       at Metsys.Bson.Serializer.WriteObject(Object document) in C:\Projects\karlseguin-Metsys.Bson-20b88f8\Metsys.Bson\Serializer.cs:line 85
       at Metsys.Bson.Serializer.WriteDocument(Object document) in C:\Projects\karlseguin-Metsys.Bson-20b88f8\Metsys.Bson\Serializer.cs:line 79
       at Metsys.Bson.Serializer.Serialize[T](T document) in C:\Projects\karlseguin-Metsys.Bson-20b88f8\Metsys.Bson\Serializer.cs:line 37
       at TestApp.Form1.button1_Click(Object sender, EventArgs e) in C:\Projects\karlseguin-Metsys.Bson-20b88f8\TestApp\Form1.cs:line 27
       at System.Windows.Forms.Control.OnClick(EventArgs e)
       at System.Windows.Forms.Button.OnClick(EventArgs e)
       at System.Windows.Forms.Button.OnMouseUp(MouseEventArgs mevent)
       at System.Windows.Forms.Control.WmMouseUp(Message& m, MouseButtons button, Int32 clicks)
       at System.Windows.Forms.Control.WndProc(Message& m)
       at System.Windows.Forms.ButtonBase.WndProc(Message& m)
       at System.Windows.Forms.Button.WndProc(Message& m)
       at System.Windows.Forms.Control.ControlNativeWindow.OnMessage(Message& m)
       at System.Windows.Forms.Control.ControlNativeWindow.WndProc(Message& m)
       at System.Windows.Forms.NativeWindow.DebuggableCallback(IntPtr hWnd, Int32 msg, IntPtr wparam, IntPtr lparam)
       at System.Windows.Forms.UnsafeNativeMethods.DispatchMessageW(MSG& msg)
       at System.Windows.Forms.Application.ComponentManager.System.Windows.Forms.UnsafeNativeMethods.IMsoComponentManager.FPushMessageLoop(Int32 dwComponentID, Int32 reason, Int32 pvLoopData)
       at System.Windows.Forms.Application.ThreadContext.RunMessageLoopInner(Int32 reason, ApplicationContext context)
       at System.Windows.Forms.Application.ThreadContext.RunMessageLoop(Int32 reason, ApplicationContext context)
       at System.Windows.Forms.Application.Run(Form mainForm)
       at TestApp.Program.Main() in C:\Projects\karlseguin-Metsys.Bson-20b88f8\TestApp\Program.cs:line 18
       at System.AppDomain._nExecuteAssembly(Assembly assembly, String[] args)
       at System.AppDomain.ExecuteAssembly(String assemblyFile, Evidence assemblySecurity, String[] args)
       at Microsoft.VisualStudio.HostingProcess.HostProc.RunUsersAssembly()
       at System.Threading.ThreadHelper.ThreadStart_Context(Object state)
       at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
       at System.Threading.ThreadHelper.ThreadStart()
  InnerException: System.ArgumentException
       Message="Error binding to target method."
       Source="mscorlib"
       StackTrace:
            at System.Delegate.CreateDelegate(Type type, MethodInfo method, Boolean throwOnBindFailure)
            at System.Delegate.CreateDelegate(Type type, MethodInfo method)
            at Metsys.Bson.MagicProperty.GetterMethod[TTarget,TParam](PropertyInfo method) in C:\Projects\karlseguin-Metsys.Bson-20b88f8\Metsys.Bson\Helpers\MagicProperty.cs:line 72
       InnerException: 

================= GenericValue class ==================

  /// <summary>

  /// Generic class to hold a value and a set/not set flag
  /// </summary>

  /// <typeparam name="T"></typeparam>
  [Serializable]
  [PersistChildren(false)]
  [ParseChildren(true)]
  [TypeConverter(typeof(SerializableExpandableObjectConverter))]
  [MergableProperty(false)]
  [Description("Generic class to hold a value and a set/not set flag")]
  public class GenericValue<T> : IXmlSerializable, IDefaultValue
  {
    #region Member data

```
private T mValue;

#endregion

#region Public Methods

/// <summary>
/// Returns a <see cref="T:System.String"/> that represents the current <see cref="T:System.Object"/>.
/// </summary>
/// <returns>
/// A <see cref="T:System.String"/> that represents the current <see cref="T:System.Object"/>.
/// </returns>
/// <filterpriority>2</filterpriority>
[Description("Returns a System.String that represents the current System.Object.")]
public override string ToString()
{
  if (!Set)
  {
    return "NotSet";
  }
  if (Equals(Value, default(T)))
  {
    return null;
  }
  return Value.ToString();
}

/// <summary>
/// Determines whether the specified <see cref="T:System.Object"/> is equal to the current <see cref="T:System.Object"/>.
/// </summary>
/// <returns>
/// true if the specified <see cref="T:System.Object"/> is equal to the current <see cref="T:System.Object"/>; otherwise, false.
/// </returns>
/// <param name="obj">The <see cref="T:System.Object"/> to compare with the current <see cref="T:System.Object"/>. </param><exception cref="T:System.NullReferenceException">The <paramref name="obj"/> parameter is null.</exception><filterpriority>2</filterpriority>
[Description(
    "Determines whether the specified System.Object is equal to the current System.Object.")]
public override bool Equals(object obj)
{
  if (ReferenceEquals(null, obj))
  {
    return false;
  }
  if (ReferenceEquals(this, obj))
  {
    return true;
  }
  if (obj.GetType() != typeof (GenericValue<T>))
  {
    return false;
  }
  return Equals((GenericValue<T>) obj);
}

/// <summary>
/// Determines whether the specified <c>GenericValue</c> is equal to the current <c>GenericValue</c>.
/// </summary>
/// <returns>
/// true if both <c>GenericValue</c> objects are not set or if their values are equal
/// </returns>
[Description(
    "Determines whether the specified GenericValue is equal to the current GenericValue.")]
public bool Equals(GenericValue<T> other)
{
  if (!Set && !other.Set)
  {
    return true;
  }
  if (Set != other.Set)
  {
    return false;
  }
  return Equals(Value, other.Value);
}

/// <summary>
/// Serves as a hash function for a particular type. 
/// </summary>
/// <returns>
/// A hash code for the current <see cref="T:System.Object"/>.
/// </returns>
/// <filterpriority>2</filterpriority>
[Description("Serves as a hash function for a particular type.")]
public override int GetHashCode()
{
  unchecked
  {
    if (Value.IsTypeDefaultValue())
    {
      return 0;
    }
    return (Value.GetHashCode()*397) ^ Set.GetHashCode();
  }
}

/// <summary>
/// Defines a method to determine if an object is a
/// default value (therefore it should not be serialized)
/// </summary>
[Description(
    "Defines a method to determine if an object is a default value (therefore it should not be serialized)"
    )]
public bool IsDefaultValue()
{
  return !Set;
}

/// <summary>
/// This method is reserved and should not be used. When implementing the <see cref="IXmlSerializable"/> interface, you should return null (Nothing in Visual Basic) from this method, and instead, if specifying a custom schema is required, apply the <see cref="T:System.Xml.Serialization.XmlSchemaProviderAttribute"/> to the class.
/// </summary>
/// <returns>
/// An <see cref="T:System.Xml.Schema.XmlSchema"/> that describes the XML representation of the object that is produced by the <see cref="M:System.Xml.Serialization.IXmlSerializable.WriteXml(System.Xml.XmlWriter)"/> method and consumed by the <see cref="M:System.Xml.Serialization.IXmlSerializable.ReadXml(System.Xml.XmlReader)"/> method.
/// </returns>
[Description(
    "This method is reserved and should not be used. When implementing the System.Xml.Serialization.XmlSchemaProviderAttribute to the class."
    )]
public XmlSchema GetSchema()
{
  return null;
}

/// <summary>
/// Generates an object from its XML representation.
/// </summary>
/// <param name="reader">The <see cref="T:System.Xml.XmlReader"/> stream from which the object is deserialized. </param>
[Description("Generates an object from its XML representation.")]
public void ReadXml(XmlReader reader)
{
  var xs = new XmlSerializer(typeof (T));
  string xml = reader.ReadInnerXml();
  if (string.IsNullOrEmpty(xml))
  {
    Set = false;
    Value = default(T);
  }
  else
  {
    var sr = new StringReader(xml);
    Value = (T) xs.Deserialize(sr);
    Set = true;
  }
}

/// <summary>
/// Converts an object into its XML representation.
/// </summary>
/// <param name="writer">The <see cref="T:System.Xml.XmlWriter"/> stream to which the object is serialized. </param>
[Description("Converts an object into its XML representation.")]
public void WriteXml(XmlWriter writer)
{
  if (!Set)
  {
    return;
  }
  var xs = new XmlSerializer(typeof (T));
  xs.Serialize(writer, Value);
}

#endregion

#region Properties

/// <summary>
/// Get or set a value indicating whether the value of this object is set
/// </summary>
[NotifyParentProperty(true)]
[DefaultValue(false)]
[Description("Get or set a value indicating whether the value of this object is set")]
public virtual bool Set
{
  get;
  set;
}

/// <summary>
/// Get or set the value of the object.
/// The Value should be ignored if the Set flag is false.
/// </summary>
[NotifyParentProperty(true)]
[Description(
    "Get or set the value of the object. The Value should be ignored if the Set flag is false.")
]
public virtual T Value
{
  get
  {
    try
    {
      bool isValueType = (typeof (T) == typeof (string)) || typeof (T).IsValueType;
      if (!isValueType && mValue.IsTypeDefaultValue())
      {
        mValue = (T) Activator.CreateInstance(typeof (T));
      }
    }
        // ReSharper disable EmptyGeneralCatchClause
    catch (Exception)
        // ReSharper restore EmptyGeneralCatchClause
    {
    }
    return mValue;
  }
  set
  {
    mValue = value;
  }
}

#endregion

/// <summary>
/// implicit operator between <c>GenericValue</c> and T
/// </summary>
/// <param name="genericValue"></param>
/// <returns></returns>
[Description("implicit operator between GenericValue and T")]
public static implicit operator T(GenericValue<T> genericValue)
{
  if (genericValue == null || !genericValue.Set)
  {
    return default(T);
  }
  return genericValue.Value;
}

/// <summary>
/// implicit operator between <c>GenericValue</c> and T
/// </summary>
/// <param name="underlyingValue"></param>
/// <returns></returns>
[Description("implicit operator between GenericValue and T")]
public static implicit operator GenericValue<T>(T underlyingValue)
{
  if (Equals(underlyingValue, default(T)))
  {
    return new GenericValue<T>();
  }
  return new GenericValue<T>
           {
               Set = true,
               Value = underlyingValue
           };
}
```

  }

================== CssClassValue class =====================

  /// <summary>

  /// A CSS class generic value object
  /// </summary>

  [Serializable]
  [Description("A CSS class generic value object")]
  public class CssClassValue : GenericValue<string>
  {
    #region Properties

```
/// <summary>
/// Get or set the value of the CSS class.
/// The Value should be ignored if the Set flag is false.
/// </summary>
[CssClassProperty]
[NotifyParentProperty(true)]
[DefaultValue("")]
[Description(
    "Get or set the value of the CSS class. The Value should be ignored if the Set flag is false."
    )]
public override string Value
{
  get
  {
    return base.Value ?? string.Empty;
  }
  set
  {
    base.Value = value;
  }
}

#endregion
```

  }

**Comment by karlseguin** [comment]
This is a bug in .NET 3.5...I hit up twitter, had a lot of people test it (using a simpler test case), and 100% of the ppl couldn't repro it in 4.0 (including me), about 75% of the people couldn't repro it in 3.5SP1, and 25% could (including me).

It specifically has to do with the reflection engine returning base generic members of overriden non-generic members...which is pretty wrong.

Here's the code that was used to test it:
http://gist.github.com/385670

**Comment by karlseguin** [comment]
I should add that I can't see a way to work around it

**Comment by Siderite** [comment]
I am sorry I didn't have the time to check out your code. All I can say is that I met with a similar issue when I was doing some old fashioned reflection and I was doing a GetProperty(name) when the name was found twice, in the base and in the inheriting class. I solved the issue back then by manually getting just the declared properties for a type, then moving on to the next and getting the first I found.

Maybe that helps.
